packages/kotti-table/src/logic/column.js

```diff
@@ -1,94 +1,101 @@
-import property from 'lodash/property'
-import negate from 'lodash/negate'
+import Vue from 'vue'
 import pick from 'lodash/pick'
-import { SORT_NONE } from '../constants'
+import { SORT_NONE, PUBLIC_COLUMN_PROPS } from '../constants'
 import { setSortedColumn } from './sort'
-import { setHiddenColumn } from './hide'
+import { setHiddenColumn, getResolvedHiddenColumns } from './hide'
 import { setFilterColumn } from './filter'
+import { resolveColumnsOrder, getOrderedColumns } from './order'
 
 export const defaultState = {
-	freshColumns: true,
-	_columns: [],
+	_destroyedColumns: {},
+	refreshColumnArray: true,
+	_columns: {},
+	_columnsArray: [],
 	columns: [],
 }
 
 export const mutations = {
-	insertColumn(store, column, index) {
-		setColumn(store.state, column, index)
-		store.commit('updateColumns')
-	},
-	swapColumns(store, fromColumn, toColumn) {
-		const { state } = store
-		const fromRealIndex = getColumnIndex(state, fromColumn)
-		const toRealIndex = getColumnIndex(state, toColumn)
-		state._columns[fromRealIndex] = toColumn
-		state._columns[toRealIndex] = fromColumn
-		store.emit(
-			'columnOrderChange',
-			state._columns.map((column, order) => {
-				column.order = order
-				return column
-			}),
-		)
-		store.commit('updateColumns')
+	insertColumn(store, { column, index }) {
+		const deleted = store.state._destroyedColumns[column.prop]
+		setColumn(store.state, { column, index, deleted })
+		deleted && (store.state._destroyedColumns[column.prop] = 0)
+		store.commit('updateColumns', { emitChange: false })
 	},
 	removeColumn(store, column) {
-		const { _columns = [] } = store.state
-		_columns.splice(_columns.indexOf(column), 1)
-		store.commit('updateColumns')
+		column = getColumn(store.state, column)
+		if (column) {
+			column._deleted = true
+			store.state._destroyedColumns[column.prop] = 1
+			store.commit('updateColumns', { emitChange: false })
+		}
 	},
-	updateColumns({ table, state }) {
+	setColumns(store, columns = store.state._columnsArray) {
+		const { state } = store
+		for (let col of columns) {
```

Comment:
```suggestion
		const pickedColumns = columns.map(column => pick(column, PUBLIC_COLUMN_PROPS)
		for (const column of pickedColumns) {
```

Reply:
Or perhaps even use a `.map().forEach()`.

Reply:
is the goal to remove the for loop ? 
cause the execution time is more or less the same

ah I see the intent is to separate the pick step from the set step

Reply:
column is not picked column, the picked columns are assigned (as in the prop values are set to the column, mutating it)
I recognise the seperation of picking and then mutating